### PR TITLE
Default compute config to HiFi3 with fp32 on WH

### PR DESCRIPTION
**Pull request by nmauriceTT**
### Ticket
#39562

### Problem description
Wormhole has a hardware in MVMUL/ELWMUL that can be triggered when using fp32 accumulation ( #38306 ). This bug happens most frequently at higher math fidelities such as HiFi4. 
As a result, fp32_acc+HiFi3 matmul/softmax can be more accurate than fp32_acc+HiFi3. 

### What's changed
- On wormhole, instead of setting default accuracy to HiFi4 + fp32_acc, set default accuracy to HiFi3 + fp32_acc. 
- If `compute_config` is passed with HiFi4 + fp32_acc then print a warning. This warning is printed once per process (c.f. `static std::atomic`). 

Updated affected ops are: matmul, softmax, minimal_matmul, dit_rms_norm_unary_fused, accumulation, reduction, batch_norm, element-wise.

The operations with updated default compute config have been identified through the following [results](https://docs.google.com/spreadsheets/d/1TEIUpGJWaq6hNKTXqRkKmJAlw5scrngZIuXH6GlMmWU/edit?usp=sharing): 

### Checklist

- [x] [Sanity tests](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:nmaurice/39562-use-hifi3-with-fp32-on-wh)
- [x] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml?query=branch:nmaurice/39562-use-hifi3-with-fp32-on-wh)
- [ ] [cpp-unit-tests](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:nmaurice/39562-use-hifi3-with-fp32-on-wh) (OK, failures in main)
- [ ] New/Existing tests provide coverage for changes
- [ ] (Optional) Ran [clang-tidy code analysis](https://github.com/tenstorrent/tt-metal/actions/workflows/code-analysis.yaml) on the PR branch to catch linting errors early

#### Model tests

If your changes cover model-related code, you should run tests corresponding to affected models and platforms (Single card, T3K, Galaxy). "Choose your pipeline" workflows facilitate running multiple kinds of tests in a single run. Each offers `models-mandatory` and `models-extended` presets.
The former includes a minimal set of tests, to be run always. The latter extends that with additional ones - use your best judgement in deciding which is the most appropriate for your PR.

- [ ] [(Single) Choose your pipeline](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:nmaurice/39562-use-hifi3-with-fp32-on-wh) (OK, failures in main)
  - [ ] `models-mandatory` preset (runs: [Device perf regressions](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) and [Frequent model and ttnn tests](https://github.com/tenstorrent/tt-metal/actions/workflows/fast-dispatch-full-regressions-and-models.yaml))
  - [ ] `models-extended` preset (runs: the mandatory tests, plus [Demo](https://github.com/tenstorrent/tt-metal/actions/workflows/single-card-demo-tests.yaml) and [Model perf](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) tests)
  - [ ] other selection - specify runs
  - [ ] Device perf regression: [OK](https://github.com/tenstorrent/tt-metal/actions/runs/23440025372) failures in [main](https://github.com/tenstorrent/tt-metal/actions/runs/23432685834/job/68163971610)

- [ ] [(T3K) Choose your pipeline](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:nmaurice/39562-use-hifi3-with-fp32-on-wh) OK, failures in main
  - [ ] `models-mandatory` preset (runs: [Unit tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-unit-tests.yaml))
  - [ ] `models-extended` preset (runs: the mandatory tests, plus [Demo](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-demo-tests.yaml) and [Model perf](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-model-perf-tests.yaml) tests)
  - [ ] other selection - specify runs

- [ ] [(Galaxy) Choose your pipeline](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:nmaurice/39562-use-hifi3-with-fp32-on-wh)
  - [ ] `models-mandatory` preset (runs: [Quick tests](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-quick.yaml))
  - [ ] `models-extended` preset (runs: the mandatory tests, plus [Demo](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-demo-tests.yaml) and [Model perf](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-perf-tests.yaml) tests)
  - [ ] other selection - specify runs

### CODE REVIEW

File: ttnn/cpp/ttnn/operations/matmul/device/matmul_device_operation.cpp [hunk 1]
```diff
@@ -1427,6 +1430,16 @@ MatmulParams create_matmul_attributes(
...
         /*default_l1_acc=*/!is_float_32);
+    // Warn if user explicitly passed HiFi4 + fp32_dest_acc_en on Wormhole B0 (hw bug #38306).
+    if (is_wormhole && parameters.compute_kernel_config.has_value() &&
+        parameters.compute_kernel_config->fp32_dest_acc_en &&
+        parameters.compute_kernel_config->math_fidelity == MathFidelity::HiFi4) {
+        log_warning(
+            tt::LogOp,
+            "On Wormhole with fp32 accumulation, output accuracy can be worse with HiFi4 than HiFi3 due to a hardware "
+            "bug. "
+            "Prefer using HiFi3 with fp32 accumulation on Wormhole.");
```
[the remark this answers is not in the document]
Comment (nmauriceTT) [reply ttnn/cpp/ttnn/operations/matmul/device/matmul_device_operation.cpp]: N/A. Wormhole B0 is the only Wormhole variants still in use.

File: ttnn/cpp/ttnn/operations/experimental/minimal_matmul/device/minimal_matmul_device_operation.cpp [hunk 2]
```diff
@@ -267,13 +268,24 @@ std::vector<Tensor> minimal_matmul(
...
         true /*packer_acc*/);
+    // Warn if user explicitly passed HiFi4 + fp32_dest_acc_en on Wormhole B0 (hw bug #38306).
+    // The default (HiFi2 + fp32_acc) is safe; only user-supplied HiFi4+fp32 is vulnerable.
+    if (arch == tt::ARCH::WORMHOLE_B0 && compute_kernel_config.has_value() && compute_kernel_config->fp32_dest_acc_en &&
+        compute_kernel_config->math_fidelity == MathFidelity::HiFi4) {
+        log_warning(
+            tt::LogOp,
+            "On Wormhole with fp32 accumulation, output accuracy can be worse with HiFi4 than HiFi3 due to a hardware "
+            "bug. "
+            "Prefer using HiFi3 with fp32 accumulation on Wormhole.");
```
[the remark this answers is not in the document]
Comment (nmauriceTT) [reply ttnn/cpp/ttnn/operations/experimental/minimal_matmul/device/minimal_matmul_device_operation.cpp]: N/A. Wormhole B0 is the only Wormhole variants still in use.

File: ttnn/cpp/ttnn/operations/reduction/generic/device/reduce_op.cpp [hunk 3]
```diff
@@ -89,12 +89,24 @@ Tensor reduce(
...
+        is_wormhole ? MathFidelity::HiFi3 : MathFidelity::HiFi4,
         /*default_approx_mode=*/false,
         /*default_fp32_acc=*/true));
+    if (is_wormhole && compute_kernel_config.has_value() && compute_kernel_config->fp32_dest_acc_en &&
+        compute_kernel_config->math_fidelity == MathFidelity::HiFi4) {
+        log_warning(
+            tt::LogOp,
+            "On Wormhole with fp32 accumulation, output accuracy can be worse with HiFi4 than HiFi3 due to a hardware "
+            "bug. "
+            "Prefer using HiFi3 with fp32 accumulation on Wormhole.");
```
[the remark this answers is not in the document]
Comment (nmauriceTT) [reply ttnn/cpp/ttnn/operations/reduction/generic/device/reduce_op.cpp]: N/A. Wormhole B0 is the only Wormhole variant still in use, there is no ambiguity here.

File: ttnn/cpp/ttnn/operations/reduction/generic/generic_reductions.cpp [hunk 4]
```diff
@@ -343,12 +343,24 @@ Tensor non_height_width_reduce(
...
+        is_wormhole ? MathFidelity::HiFi3 : MathFidelity::HiFi4,
         /*default_approx_mode=*/false,
         /*default_fp32_acc=*/true));
+    if (is_wormhole && compute_kernel_config.has_value() && compute_kernel_config->fp32_dest_acc_en &&
+        compute_kernel_config->math_fidelity == MathFidelity::HiFi4) {
+        log_warning(
+            tt::LogOp,
+            "On Wormhole with fp32 accumulation, output accuracy can be worse with HiFi4 than HiFi3 due to a hardware "
+            "bug. "
+            "Prefer using HiFi3 with fp32 accumulation on Wormhole.");
```
[the remark this answers is not in the document]
Comment (nmauriceTT) [reply ttnn/cpp/ttnn/operations/reduction/generic/generic_reductions.cpp]: N/A. Wormhole B0 is the only Wormhole variants still in use.

File: ttnn/cpp/ttnn/operations/experimental/transformer/dit_rms_norm_unary_fused/dit_rms_norm_unary_fused.cpp [hunk 5]
```diff
@@ -28,13 +28,28 @@ ttnn::Tensor dit_rms_norm_unary_fused(
...
     }
 
+    // Warn if user explicitly passed HiFi4 + fp32_dest_acc_en on Wormhole B0 (hw bug #38306).
+    if (is_wormhole && compute_kernel_config.has_value() && compute_kernel_config->fp32_dest_acc_en &&
+        compute_kernel_config->math_fidelity == MathFidelity::HiFi4) {
+        log_warning(
+            tt::LogOp,
+            "On Wormhole with fp32 accumulation, output accuracy can be worse with HiFi4 than HiFi3 due to a hardware "
+            "bug. "
+            "Prefer using HiFi3 with fp32 accumulation on Wormhole.");
```
[the remark this answers is not in the document]
Comment (nmauriceTT) [reply ttnn/cpp/ttnn/operations/experimental/transformer/dit_rms_norm_unary_fused/dit_rms_norm_unary_fused.cpp]: N/A. Wormhole B0 is the only Wormhole variants still in use.

File: ttnn/cpp/ttnn/operations/normalization/softmax/device/softmax_device_operation.cpp [hunk 6]
```diff
@@ -401,6 +401,23 @@ SoftmaxDeviceOperation::create_op_performance_model(
...
+    tt::ARCH arch, const std::optional<const DeviceComputeKernelConfig>& compute_kernel_config, bool is_fp32) {
+    const auto is_wormhole = arch == tt::ARCH::WORMHOLE_B0;
+    const auto default_fidelity = (is_wormhole && is_fp32) ? MathFidelity::HiFi3 : MathFidelity::HiFi4;
+    if (is_wormhole && compute_kernel_config.has_value() && compute_kernel_config->fp32_dest_acc_en &&
+        compute_kernel_config->math_fidelity == MathFidelity::HiFi4) {
+        log_warning(
+            tt::LogOp,
+            "On Wormhole with fp32 accumulation, output accuracy can be worse with HiFi4 than HiFi3 due to a hardware "
+            "bug. "
+            "Prefer using HiFi3 with fp32 accumulation on Wormhole.");
```
[the remark this answers is not in the document]
Comment (nmauriceTT) [reply ttnn/cpp/ttnn/operations/normalization/softmax/device/softmax_device_operation.cpp]: N/A. Wormhole B0 is the only Wormhole variant still in use.

File: ttnn/cpp/ttnn/operations/matmul/device/matmul_device_operation.cpp [hunk 7]
```diff
@@ -1381,7 +1381,10 @@ MatmulParams create_matmul_attributes(
     const auto increase_fidelity = !has_program_config && !has_user_grid && !are_inputs_low_precision_df;
     auto math_fidelity = increase_fidelity ? MathFidelity::HiFi2 : MathFidelity::LoFi;
     bool are_inputs_32F = (input_tensor_a.dtype() == DataType::FLOAT32 && input_tensor_b.dtype() == DataType::FLOAT32);
-    math_fidelity = are_inputs_32F ? MathFidelity::HiFi4 : math_fidelity;
+    // Due to hardware bug (#38306), HiFi4 + fp32_dest_acc_en produces incorrect results on Wormhole B0.
```
[the remark this answers is not in the document]
Comment (nmauriceTT) [reply ttnn/cpp/ttnn/operations/matmul/device/matmul_device_operation.cpp]: #38306 contains issue with information regarding the hardware bug and is the most correct link in this context.

File: ttnn/cpp/ttnn/operations/eltwise/unary/device/unary_program_factory.cpp [hunk 8]
```diff
@@ -154,12 +154,17 @@ UnaryProgramFactory::cached_program_t UnaryProgramFactory::create(
 
     auto path = fmt::format("{}/{}", compute_root, utils::get_compute_kernel_path(ops_chain[0].type(), input.dtype()));
 
+    // Due to hardware bug (#38306), HiFi4 + fp32_dest_acc_en produces incorrect results on Wormhole B0.
+    // Use HiFi3 when fp32_dest_acc_en is True on Wormhole B0.
+    const auto default_fp32_acc_math_fidelity =
+        (args.fp32_dest_acc_en && device->arch() == tt::ARCH::WORMHOLE_B0) ? MathFidelity::HiFi3 : MathFidelity::HiFi4;
+
```
[the remark this answers is not in the document]
Comment (nmauriceTT) [reply ttnn/cpp/ttnn/operations/eltwise/unary/device/unary_program_factory.cpp]: Updated PR description.

File: ttnn/cpp/ttnn/operations/reduction/generic/generic_reductions.cpp [hunk 9]
```diff
@@ -343,12 +343,24 @@ Tensor non_height_width_reduce(
...
         std::nullopt,
-        MathFidelity::HiFi4,
+        is_wormhole ? MathFidelity::HiFi3 : MathFidelity::HiFi4,
         /*default_approx_mode=*/false,
         /*default_fp32_acc=*/true));
+    if (is_wormhole && compute_kernel_config.has_value() && compute_kernel_config->fp32_dest_acc_en &&
+        compute_kernel_config->math_fidelity == MathFidelity::HiFi4) {
+        log_warning(
+            tt::LogOp,
+            "On Wormhole with fp32 accumulation, output accuracy can be worse with HiFi4 than HiFi3 due to a hardware "
```
[the remark this answers is not in the document]
Comment (nmauriceTT) [reply ttnn/cpp/ttnn/operations/reduction/generic/generic_reductions.cpp]: Added tt-logger.hpp include.

**Review by bbradelTT — APPROVED** [review]

File: ttnn/cpp/ttnn/operations/core/compute_kernel/compute_kernel_config.cpp [hunk 10]
```diff
@@ -38,6 +41,32 @@ DeviceComputeKernelConfig init_device_compute_kernel_config(
...
+    }
+    if (!user_compute_kernel_config.has_value()) {
+        return;
+    }
+    const auto& cfg = user_compute_kernel_config.value();
+    if (!cfg.fp32_dest_acc_en || cfg.math_fidelity != MathFidelity::HiFi4) {
+        return;
+    }
+    // Only print warning once per process.
+    static std::atomic<bool> warning_generated{false};
```
Comment (iwroszTT) [inline ttnn/cpp/ttnn/operations/core/compute_kernel/compute_kernel_config.cpp:60]: Do we use an std::atomic to print only once even with multiple threads?

Comment (iwroszTT) [reply ttnn/cpp/ttnn/operations/core/compute_kernel/compute_kernel_config.cpp:60]: If that's the case, I believe this would be a case of time-of-check-time-of-use scenario. What could be used instead is a compare exchange atomic operation.

If it's a single thread context, then perhaps a simple static variable would be enough?

**Review by iwroszTT — APPROVED** [review]

File: ttnn/cpp/ttnn/operations/core/compute_kernel/compute_kernel_config.cpp (hunk 10 above)
Comment (bbradelTT) [reply ttnn/cpp/ttnn/operations/core/compute_kernel/compute_kernel_config.cpp:60]: This is just to prevent flooding the CI logs for models. Printing a few times is okay.

Comment (nmauriceTT) [reply ttnn/cpp/ttnn/operations/core/compute_kernel/compute_kernel_config.cpp:60]: @iwroszTT Yes, the goal is to only print once per process. Without it, we would get a good amount of spam in [model logs](https://github.com/tenstorrent/tt-metal/actions/runs/23466399746/job/68341550021#step:8:169).

I'm not sure how much multithreading support we have in ttnn (either python or C++ side). I saw code with `thread_local static` (would work on a per thread basis), but `std::atomic` would make it only print once per process. 
I can change it to a CAS (I went for the easiest-to-read approach; assuming that the atomic operation is not a performance bottleneck).

**Review by mgajewskiTT — APPROVED** [review]

File: ttnn/cpp/ttnn/operations/core/compute_kernel/compute_kernel_config.cpp (hunk 10 above)
Comment (nmauriceTT) [reply ttnn/cpp/ttnn/operations/core/compute_kernel/compute_kernel_config.cpp:60]: @iwroszTT I've changed the atomic code to use `compare_exchange_strong`

**Review by mateusznowakTT — APPROVED** [review]

**Review by cglagovichTT — APPROVED** [review]
